dns/provider.go

```diff
@@ -277,11 +319,16 @@ Retry:
 			retry_tcp = true
 		}
 
+		// Reset retries counter on protocol change
+		retries = meta.(*DNSClient).retries
 		goto Retry
 	case nil:
-		fallthrough
-	default:
 		// do nothing
+	default:
+		if isTimeout(err) && retries > 0 {
+			retries--
+			goto Retry
```

Comment:
Originally I had the `Retry` happen here as well, but from my testing and understanding of the code it didn't retry a failed *connection*, it would only retry a failed action occurring on that connection, which is what lead to the i/o timeout. I may be wrong and have simply implemented this wrong, I'll give your PR a spin with `-parallelism=100` on a couple hundred records and see if I run into the same issues as before #67.

Reply:
There shouldn't be any difference, the timeout is implemented entirely within the DNS client not the TF provider so it shouldn't matter where the retry is in the provider code. This code is detecting if the error object is thrown by the underlying `net.Dialer` as its error objects implement a `Timeout()` method so it just checks the error conforms to that interface and that the `Timeout()` method returns `true`. Any other error from either the `net.Dialer` or  the DNS client doesn't implement that interface or returns `false`, such as a connection refused, or a DNS refused (such as not setting the TSIG key correctly).

Reply:
So I integrated your changes into my branch using this commit: https://github.com/DeviaVir/terraform-provider-dns/commit/4a7a5022378bc422965307e63d3db2f2e5d02812
the difference between yours and mine is that I updated `miekg/dns` to latest and `ErrTruncated` was removed (see https://github.com/miekg/dns/commit/1ff265a78454967a4321dc2c0e38e267323bc3d4) so I rewrote the `switch` statement slightly.

I do have the `SERVFAIL` errors happening again (with `-parallelism=100`) which I'd love to see retried for at least 3 times as well, but I can understand the viewpoint from the provider as the server not handling that high parallelism from a single client as valid.

Reply:
I'd like to propose this commit to retry the `SERVFAIL`:
https://github.com/DeviaVir/terraform-provider-dns/commit/9ff9dd2d6576d2267db3ca354550e462246cc9de

thoughts?

Reply:
I'm in two minds that ramping up the parallelism to the point the server doesn't cope isn't something the provider should handle; presumably there is a `-parallelism` value that works 100% for you? (Also, what DNS server are you using? Is it possible you're hitting some sort of DoS ratelimit in it?)

However, client resolvers usually retry against different DNS servers upon receipt of `SERVFAIL` responses but that's not possible here as we configure one explicit server, so maybe it's acceptable. If you have misconfigured TSIG details for example, then it should still fail fast with the appropriate error code.

I'm not going to muddy this PR with updating vendor libraries, (was there any particular reason to update?), but thanks for the heads-up, I'll see if that change can be made regardless. However your error checking is still wrong here; you're retrying with a different transport on any error (including I/O timeouts) instead of just the truncation case.

Reply:
> Also, what DNS server are you using? Is it possible you're hitting some sort of DoS ratelimit in it?

Quite possibly, it's a "standard" bind server with no particularly interesting config.

> was there any particular reason to update?

That was mostly out of hope that it would help resolve the problem. It didn't, and I don't think there's some feature/bugfix that is worth taking over at this time.

I landed on [this](https://github.com/DeviaVir/terraform-provider-dns/commit/a36a827b15524238720f1b11b55e4e71cc653e77) successfully working with UDP as the transport and that high parallelism.

Reply:
In theory a single request can be unlucky and fail N times in a row. You can always increase the number of retries now that it's configurable. Interesting that it's BIND, I would have thought it would've been one of the more stable, possibly worth an investigation.

I've updated the PR so it now also retries on `SERVFAIL`. The truncation handling will have to be adapted if/when the DNS library is updated. There is an acceptance test that covers that particular case anyway.

Reply:
Yeah, I'm having success with `retries = 100`, but it surely does look off. I think your PR looks fantastic right now.